### ISSUE TITLE
Stop relying on auto-generated names in stdlib

### ISSUE THREAD

**Issue by jashug** [coq/coq#6781]
Proofs in the stdlib currently rely on auto-generated names in lots of places (one example I was looking at today while attempting #6607, [here](https://github.com/coq/coq/blob/4d8903c06fd9fd2aca793da8bb55448906347a8c/theories/Reals/RList.v#L49) the `H` on the second line is not introduced, but this is a really pervasive phenomenon). As suggested in #2301, I believe that this makes proofs harder to maintain, and makes it more difficult to change other parts of the system.

I have specifically run into issues with the current state when trying #6607 and when trying to simplify the `intuition` tactic.

Working on top of #6582, I would like to edit the proofs in the standard library so that it compiles with the `-mangle-names "_"` option.

I see this providing several benefits:
* A stdlib more robust to changes in the rest of Coq
* Experience writing proofs in this style, which may provide motivation to make the process easier
* Assuming we also consider the robust style good practice, leading by example

A possible problem with this is that the test suite becomes less sensitive to changes in generated names, making it easier for changes to break users not writing robust proofs (especially unnoticed by developers).

As long as the changes are confined to Opaque proofs, or in the case of transparent terms built with tactics limited to providing explicit names, I don't see any potential for compatibility issues.

This process should produce incremental benefits, so even if I run out of steam partway through there is no wasted work.

I have heard rumors of work on a stdlib2 by @maximedenes in the future. However, I do not believe there are plans for the current stdlib to disappear in say the next year given how much work is dependent on it. Therefore I hope this has benefit independent of stdlib2.

I am volunteering to edit the proofs (in what free time I have), but I would need someone to review the pull requests I generate. Is there someone who is willing and able to do this? (@herbelin? @ppedrot? Who is responsible for maintaining the standard library?)

**Comment by herbelin** [comment]
> I have specifically run into issues with the current state when trying #6607 and when trying to simplify the intuition tactic.

Indeed, and the `Reals` library is particularly good at using generated names...

About stdlib2, you should ask for more details but what I understood is that it was primarily about naming conventions for lemmas and definitions. So, I (a priori) believe that it is orthogonal to what you are proposing.

> I am volunteering to edit the proofs (in what free time I have), but I would need someone to review the pull requests I generate. Is there someone who is willing and able to do this? (@herbelin? @ppedrot?)

I understand correctly, it is only about changing names within proofs, so it should not have a big impact observationally speaking, right? So, I can help @Zimmi48 if ever needed.

>  Who is responsible for maintaining the standard library?

It depends on the parts of the library. You can see the list [there](https://github.com/coq/coq/wiki/Maintainers).

About tactics like `intuition`, there is the long-standing wish of supporting an `as` clause for it (or even to provide a generic `as` clause working for all kind of arbitrary tactic), but it requires a bit of expertise with the proof engine.

**Comment by Zimmi48** [comment]
> About stdlib2, you should ask for more details but what I understood is that it was primarily about naming conventions for lemmas and definitions. So, I (a priori) believe that it is orthogonal to what you are proposing.

Not quite, this is (AFAIU) about removing some duplication between the standard library and some popular external libraries like math-comp on the aspects that could be shared (like the duality Z vs int). It is also (AFAIU) about promoting a general style (and this include naming conventions, the aspect on which you focused, but this includes as well not relying on auto-generated names, using bullets or not, etc).

> If I understand correctly, it is only about changing names within proofs, so it should not have a big impact observationally speaking, right?

Indeed, it shouldn't have as long as it is limited to opaque lemmas.

**Comment by SkySkimmer** [comment]
Even with transparent lemmas alpha renaming should rarely break other 
lemmas no?

Gaëtan Gilbert

**Comment by Zimmi48** [comment]
Yes, and the tests would catch the problems anyways.

**Comment by jashug** [comment]
This work is currently blocked on #6785.

**Comment by jashug** [comment]
Recording for my future self: compile with `COQUSERFLAGS="-mangle-names _" make`.

After #6785, the next blocker is #12763.

**Comment by RalfJung** [comment]
Here's another name mangling issue: `abstract` breaks. https://github.com/coq/coq/issues/12928

**Comment by RalfJung** [comment]
And another one: strange "name already used" errors. https://github.com/coq/coq/issues/12944

**Comment by RalfJung** [comment]
But even with that issue, [all of std++ can now be made to build with name mangling enabled](https://gitlab.mpi-sws.org/iris/stdpp/-/merge_requests/182). :-)

**Comment by RalfJung** [comment]
Here's another name mangling problem: https://github.com/coq/coq/issues/13129

**Comment by RalfJung** [comment]
One issue in `setoid_rewrite` was fixed, but others remain: https://github.com/coq/coq/issues/13246

**Comment by RalfJung** [comment]
More of Iris now builds with name mangling enabled, but there are still issues left: https://github.com/coq/coq/issues/13413

**Comment by jashug** [comment]
A problem with `funind` and name mangling that shows up in the standard library: #13729.

**Comment by RalfJung** [comment]
In Iris we're running into many instances of https://github.com/coq/coq/issues/13759; isn't that also a problem for the stdlib?

**Comment by RalfJung** [comment]
The good news is that if I work around https://github.com/coq/coq/issues/13759, then all of Iris builds with name mangling enabled. :) So after many rounds of fixing bugs in this infrastructure, it now seems to be in a pretty good state. :D

**Comment by SkySkimmer** [comment]
#13759 finally closed!
